src/vendor/vendor.c

```diff
@@ -1834,39 +1809,46 @@ int DiscoverDM_ForAllComponents(char *objs_filename, char *params_filename)
 ** \return  USP_ERR_OK if successful
 **
 **************************************************************************/
-int DiscoverDM_ForComponent(rdk_component_t *rdkc, kv_vector_t *rdk_objects, kv_vector_t *rdk_params)
+int DiscoverDM_ForComponent(kv_vector_t *rdk_objects, kv_vector_t *rdk_params)
 {
-    int i;
-    int ccsp_err;
+    int rbus_err;
     int err;
-    parameterInfoStruct_t **param_infos = NULL;
-    parameterInfoStruct_t *param_info;
-    int num_param_infos = 0;
     char *write_status;
+    rbusElementInfo_t* elem;
+    const char *component, *c1;
 
     // Exit if unable to determine all instances provided by this rdk component
-    USP_LOG_Info("%s: Getting parameters and objects for '%s'", __FUNCTION__, rdkc->group_name);
-    ccsp_err = CcspBaseIf_getParameterNames(bus_handle, rdkc->component_name, rdkc->dbus_path, "Device.", 0, 
-                                          &num_param_infos, &param_infos);
-    if (ccsp_err != CCSP_SUCCESS)
+    USP_LOG_Info("%s: Getting parameters and objects for CCSP components using rbus", __FUNCTION__);
+    rbusElementInfo_t* elems = NULL;
+    rbus_err = rbusElementInfo_get(bus_handle,"Device.", RBUS_MAX_NAME_DEPTH, &elems);
+    if (rbus_err != RBUS_ERROR_SUCCESS)
     {
-        USP_LOG_Error("%s: CcspBaseIf_getParameterNames(%s) failed (%d- %s)", __FUNCTION__, rdkc->group_name, ccsp_err, ToCcspErrString(ccsp_err));
-        return USP_ERR_INTERNAL_ERROR;
+       USP_LOG_Error("%s: rbusElementInfo_get(%s) failed (%d- %s)", __FUNCTION__, elems->name, rbus_err, ToRbusErrString(rbus_err));
+       return USP_ERR_INTERNAL_ERROR;
     }
 
     // Iterate over all parameters and objects found
-    for (i=0; i<num_param_infos; i++)
-    {
-        param_info = param_infos[i];
-        write_status = (param_info->writable) ? "RW" : "RO";
-        Add_NameToDM(rdkc, param_info->parameterName, write_status, rdk_objects, rdk_params);
+    elem = elems;
+    component = NULL;
+    char* component_name = NULL;
+    while(elem)
+    {
+        component = elem->component;
+	c1 = NULL;
+        write_status = (elem->type == RBUS_ELEMENT_TYPE_TABLE ? elem->access & RBUS_ACCESS_ADDROW : elem->access & RBUS_ACCESS_SET)  ? "RW" : "RO";
```

Comment:
Given that R-Bus APIs give you the type of the element, you should switch on that and then directly call Add_ObjectToDM() or Add_ParamToDM(). I don't think that there's a need for Add_NameToDM() anymore.

Reply:
calling either Add_ObjectToDM() or Add_ParamToDM() causing below issues
-> calling  Add_ParamToDM() alone
```
DM_PRIV_AddSchemaPath: Device.X_RDK_Ethernet.Link.{i} must be registered before Device.X_RDK_Ethernet.Link.1.Enable
USP Agent aborted unexpectedly
```
-> calling Add_ObjectToDM() alone
`DM_PRIV_GetNodeFromPath: Path is invalid: Device.DeviceInfo.X_RDKCENTRAL-COM_LastRebootReason`